app/models/topic.rb

```diff
@@ -1632,6 +1632,36 @@ def access_topic_via_group
       .first
   end
 
+  def incoming_email_addresses(group: nil, received_before: Time.zone.now)
+    email_addresses = group.present? ? Set[group.email_username] : Set.new
+
+    # TODO(martin) Look at improving this N1, it will just get slower the
+    # more replies/incoming emails there are for the topic.
+    self.incoming_email.where("created_at <= ?", received_before).each do |incoming_email|
+      to_addresses = incoming_email.to_addresses_split
+      cc_addresses = incoming_email.cc_addresses_split
+      combined_addresses = [to_addresses, cc_addresses].flatten
+
+      # We only care about the emails addressed to the group or CC'd to the
+      # group if the group is present.
+      if group.present? && combined_addresses.any?
+        next if combined_addresses.none? { |address| address =~ group.email_username_regex }
+      end
+
+      email_addresses.add(incoming_email.from_address)
+      email_addresses.merge(to_addresses) if to_addresses.present?
+      email_addresses.merge(cc_addresses) if cc_addresses.present?
+    end
+
+    email_addresses.subtract([nil, ''])
+
+    if group.present?
```

Comment:
Nitpick: You can make this a bit shorter.

```suggestion
    email_addresses.delete(group.email_username) if group.present?
```

Reply:
Ah nice yeah the delete is better but won’t need it anyway if we don’t put that email in the set to begin with.

Reply:
I think IncomingEmail records with `from_address == group.email_username` can exist, right? We [create an IncomingEmail](https://github.com/discourse/discourse/blob/64ba5b1d21f9c5c24a3799a5615dd321c6f2e778/app/jobs/regular/group_smtp_email.rb#L35-L45) record to avoid double sync of emails sent using SMTP.

Reply:
Oh right of course, nvm 😅